Feature/#14 스터디 crud 기능 구현

lcqff opened the pull request:
## #️⃣ 연관된 이슈
close: #14

## 📝 작업 내용

스터디 CRUD를 작업했습니다. 
원래 스터디장 위임도 제가 하기로 했었는데 저희가 study에 leader_Id를 빼버리고 Role 테이블을 아직 안만들어서 아직 못했습니다
밥먹고 와서 코드 한 번 더 보면서 수정할 부분 있으면 수정하겠습니다~!

## 💬 리뷰 요구사항(선택)
- 스터디 조회 테스트) 만들어진 api 문서의 스터디 조회에 responseBody가 안뜨는데 이유를 모르겠습니다...ㅡㅡ;;
- 스터디 수정) 수정해도 될 부분은 이름,설명,시작 날짜, 종료 날짜, 상태 정도로 보여서 수정 dto에 해당 필드만 만들어뒀습니다. 그런데 사용자가 위 필드 중 일부만 수정했을 경우 이전 데이터에서 변경되지 않은 부분의 데이터를 들고오는 작업은 프론트에서 해야하는 건가요 백엔드에서 해야하는 건가요...? 백엔드에서 하려니 dto 필드값에 @NotNull 어노테이션을 지워야 할 거 같아 일단 그대로 뒀습니다.


## ⏰ 예상했던 소요시간/실제 소요시간(선택)

예상했던 것보다 정확히 세배 걸린듯요..하하

lcqff committed:
feat: 스터디 CRUD 작성

- 스터디 생성
- 스터디 삭제
- 스터디 수정
- 스터디 종료
- 스터디 조회

#14

lcqff committed:
test: 통합/문서 테스트

- api 테스트
- 서비스 테스트
- 도메인 테스트

#14

lcqff committed:
test: api 문서 테스트

#14

amaran-th commented on src/main/java/doore/study/api/StudyController.java:
```diff
@@ -0,0 +1,53 @@
+@RestController
+@RequiredArgsConstructor
+public class StudyController {
+    private final StudyService studyService;
```
저희 StudyService를 StudyCommandService와 StudyQueryService로 분리하기로 해서 반영해주시면 좋을 것 같아요~~!

amaran-th commented on src/main/java/doore/study/api/StudyController.java:
```diff
@@ -0,0 +1,53 @@
+        studyService.updateStudy(studyUpdateRequest, studyId);
+    }
+
+    @PostMapping("/studies/{studyId}/termination")
```
Study의 정보 중 하나를 수정하는 API이니 Post보다는 Patch가 더 적합하지 않나 싶습니다!

amaran-th commented on src/main/java/doore/study/api/StudyController.java:
```diff
@@ -0,0 +1,53 @@
+        return studyService.findStudyById(studyId);
+    }
+
+    @PatchMapping("/studies/{studyId}")
```
'스터디 상세 정보'라는 하나의 집합을 통째로 수정하는 api이니 Patch보다는 Put이 더 적합하지 않을까요??
멱등성을 만족하고 있기도 하구요~

lcqff replied:
HTTP 메서드의 단위?로 dto도 고려해야하는건가욥? 저는 스터디의 정보 중 일부를 제외한 정보만 수정하니 일부(이름, 소개, 시작날짜, 종료날짜)를 수정하는 거라 생각했습니다!

shkisme replied:
> 스터디 수정) 수정해도 될 부분은 이름,설명,시작 날짜, 종료 날짜, 상태 정도로 보여서 수정 dto에 해당 필드만 만들어뒀습니다. 그런데 사용자가 위 필드 중 일부만 수정했을 경우 이전 데이터에서 변경되지 않은 부분의 데이터를 들고오는 작업은 프론트에서 해야하는 건가요 백엔드에서 해야하는 건가요...? 백엔드에서 하려니 dto 필드값에 NotNull 어노테이션을 지워야 할 거 같아 일단 그대로 뒀습니다.

프론트에서 스터디 상세 정보 조회 데이터를 들고와서 스터디 수정 폼에 넣음 -> 클라이언트가 바꾸고 싶은 부분 변경함 -> 백엔드에 보내주는 방식으로 보통 동작하는 것으로 알고 있습니다!
그래서 API는 일부가 아닌 스터디 전체 수정 API를 만들어두면 될 것 같고(현재 만들어주신 것 처럼), 따라서 HTTP 메서드는 Patch보다는 Put이 어울린다고 생각합니다.

lcqff replied:
오옹 사용자가 일부만 수정했더라도 결국 request되는 데이터는 전체니까 그런거군요~! 알겠습니다!

amaran-th commented on src/main/java/doore/study/domain/CurriculumItem.java:
```diff
@@ -10,13 +10,17 @@
 
 @Entity
 @Getter
+@Setter
```
롬복을 사용한 Setter 구현은 엔티티의 모든 필드를 외부에서 변경할 수 있게 열어두는 거라 위험합니다...!
앞서 언급했듯 setter보다는 변경 의도가 담긴 이름으로 업데이트 메서드를 구현해주시면 좋을 것 같아요.

amaran-th commented on src/main/java/doore/study/application/StudyService.java:
```diff
@@ -0,0 +1,58 @@
+    private final CurriculumItemRepository curriculumItemRepository;
+
+    public void createStudy(final StudyCreateRequest request, final Long teamId) {
+        //todo: 존재하는 팀인지 유효성 검사
```
혹시 이 부분은 왜 구현되지 않은 것인지 알 수 있을까요?

lcqff replied:
teamService에 해당 부분이 아직 없길래 그냥 넘겼는데 생각해보니 teamRepository로 하면 되는군요~ 수정해뒀습니다!

amaran-th commented on src/main/java/doore/study/application/StudyService.java:
```diff
@@ -0,0 +1,58 @@
+        if (study.getStatus() == ENDED) {
+            throw new StudyException(TERMINATED_STUDY);
+        }
+        study.setStatus(ENDED);
```
setter는 책임주도 설계를 해친다는 이유로 지양되는 메서드입니다. 
그래서 저는 불가피하게 setter가 필요한 경우, set~이라는 이름보다는 update~와 같이 변경 의도가 드러나는 이름으로 네이밍하는 편이에요.
그리고 이 로직의 경우, 스터디의 상태를 Ended로 변경한다는 책임이 Study가 아니라 Service에 있는 것 같습니다. study.terminate()와 같은 메서드로 구현해주시면 캡슐화 측면에서도 좋을 것 같아요~

lcqff replied:
오옹 넘 잘 배워갑니다~!

amaran-th commented on src/main/java/doore/study/application/StudyService.java:
```diff
@@ -0,0 +1,58 @@
+
+    public StudyDetailResponse terminateStudy(Long studyId) {
+        Study study = studyRepository.findById(studyId).orElseThrow(() -> new StudyException(NOT_FOUND_STUDY));
+        if (study.getStatus() == ENDED) {
```
이 조건문은 study 안으로 넣을 수 있겠네요~
예) study.isEnded()

amaran-th commented on src/main/java/doore/study/application/StudyService.java:
```diff
@@ -0,0 +1,58 @@
+
+    public void updateStudy(StudyUpdateRequest request, Long studyId) {
+        Study study = studyRepository.findById(studyId).orElseThrow(() -> new StudyException(NOT_FOUND_STUDY));
+        study.update(request);
```
dto가 도메인 수준까지 내려오는 게 괜찮은 것인지 고민해보면 좋을 것 같아요

amaran-th commented on src/main/java/doore/study/application/dto/request/StudyCreateRequest.java:
```diff
@@ -0,0 +1,76 @@
+    @Builder
+    public StudyCreateRequest(String name, String description, LocalDate startDate, LocalDate endDate,
+                              StudyStatus status,
+                              Boolean isDeleted, Long cropId, List<CurriculumItemsRequest> curriculumItems) {
```
status 다음의 개행을 없애주시면 보기 더 깔끔할 것 같아요!

amaran-th commented on src/main/java/doore/study/application/dto/response/StudyDetailResponse.java:
```diff
@@ -0,0 +1,61 @@
+    public static StudyDetailResponse from(Study study) {
+        List<CurriculumItem> curriculumItems = study.getCurriculumItems();
+        List<CurriculumItemResponse> curriculumItemResponses = curriculumItems == null ? Collections.emptyList()
+                : curriculumItems.stream().map(CurriculumItemResponse::from)
```
한 라인에 .은 한 번만 찍도록 하면 더 가독성이 좋아질 것 같습니다~!

amaran-th commented on src/main/java/doore/study/domain/Study.java:
```diff
@@ -10,15 +11,20 @@
 
 @Entity
 @Getter
+@Setter
```
여기도 마찬가지로 setter 사용 지양해주시면 좋을 것 같습니다~

amaran-th commented on src/main/java/doore/study/exception/StudyExceptionType.java:
```diff
@@ -0,0 +1,26 @@
+
+public enum StudyExceptionType implements BaseExceptionType {
+    NOT_FOUND_STUDY(HttpStatus.NOT_FOUND, "스터디를 찾을 수 없습니다."),
+    TERMINATED_STUDY(HttpStatus.FORBIDDEN, "이미 종료된 스터디입니다.");
```
이 다음에 개행이 들어가면 더 보기 좋을 것 같아요~!

amaran-th replied:
그리고 enum 값 이름과 메시지에 상황 설명이 들어가면 어떤 상황에서 해당 에러가 발생하는 지 유추하기 쉬울 것 같습니다!
```java
ALREADY_TERMINATED_STUDY(HttpStatus.FORBIDDEN, "이미 종료된 스터디입니다.");
```

amaran-th commented on src/test/java/doore/restdocs/docs/StudyApiDocsTest.java:
```diff
@@ -0,0 +1,97 @@
+    @Test
+    @DisplayName("스터디를 생성한다.")
+    public void 스터디를_생성한다() throws Exception {
+        StudyCreateRequest request = studyCreateRequest();
```
저희 문서화 테스트는 모킹하기로 했었는데 그렇게 해주시면 감사하겠습니다~!

lcqff replied:
다른 테스트들은 모르겠는데 생성 테스트도 모킹으로 테스트가 가능한가요? 필드값 없어도 되나요? 
일단 다른 테스트들은 수정해두었습니다!

amaran-th replied:
넵 가능합니다! 생성 테스트도 모킹 부탁드려요~

amaran-th reviewed (changes requested):
한 번에 많은 기능을 구현하시느라 수고 많으셨습니다!
제 의견을 (좀 많이) 남겼으니 확인해주시면 감사하겠습니다~!

lcqff committed:
refactor: dto 클래스를 record로

#14

lcqff committed:
fix: class를 record로 바꾸며 생긴 getter 문제 해결

#14

lcqff committed:
refactor: 수정 기능에서의 save 코드 삭제

#14

lcqff committed:
refactor: 필요없는 저장 코드 삭제

#14

lcqff committed:
refactor: 스터디 종료 Post -> Patch

#14

lcqff committed:
refactor: 스터디 생성시 팀 존재 유효성 검사 추가

#14

lcqff committed:
chore: 불필요한 어노테이션 제거

#14

lcqff committed:
refactor: domain에 dto를 전달하지 않는다.
- study 도메인에 dto 제거

#14

lcqff committed:
refactor: 조건문을 도메인 안에 삽입한다.

- IsEnded 조건문을 Study 내에 추가

#14

lcqff committed:
refactor: 엔티티의 변경 책임을 엔티티에게 둔다.

- Study의 상태 변경 책임을 Service에서 제거

#14

lcqff committed:
chore: 코드 가독성 수정

#14

lcqff committed:
refactor: CurriculumItems setter 삭제

#14

lcqff committed:
refactor: 문서 테스트에서 모킹 객체를 사용한다.

#14

lcqff committed:
refactor: service를 QuertService와 CommandService로 분리한다.

#14

lcqff committed:
refactor: 스터디 생성시 status와 isDeleted 초기값 설정

- status는 UPCOMING, isDeleted는 false로 초기값 설정해줌

#14

lcqff committed:
refactor: 예외처리 이름 변경

- TERMINATE_STUDY -> ALREADY_TERMINATE_STUDY

#14

shkisme commented on src/main/java/doore/study/api/StudyController.java:
```diff
@@ -0,0 +1,53 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequiredArgsConstructor
```
클래스 단에 `@Validated` 붙여줘야 `@Valid`가 동작합니다!

shkisme commented on src/main/java/doore/study/api/StudyController.java:
```diff
@@ -0,0 +1,55 @@
+    @ResponseStatus(HttpStatus.CREATED)
+    public void createStudy(@Valid @RequestBody StudyCreateRequest studyRequest, @PathVariable Long teamId) {
+        studyCommandService.createStudy(studyRequest, teamId);
+    }
```
통일성 있게 ResponseEntity를 반환해주면 좋을 것 같아요. https://github.com/BDD-CLUB/01-doo-re-back/pull/32#discussion_r1464811019

shkisme commented on src/main/java/doore/study/api/StudyController.java:
```diff
@@ -0,0 +1,55 @@
+        studyCommandService.updateStudy(studyUpdateRequest, studyId);
+    }
+
+    @PatchMapping("/studies/{studyId}/termination")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void terminateStudy(@PathVariable Long studyId) {
+        studyCommandService.terminateStudy(studyId);
+    }
```
스터디 종료 api만 있나용? 스터디 상태 변경 api로 스터디 상태를 `진행 전`, `진행 중`, `종료` 3가지로 바꿀 수 있으면 좋을 것 같아요

lcqff replied:
저도 오늘 그 기능 추가하려다가 그냥 스터디 수정에서 상태 수정하게 만드는게 낫나?(지금처럼) 싶어서 관뒀는데 또 생각해보니 상태 변경 api가 있는 편이 나을 거 같고...🤔 스터디장이 보는 스터디 페이지 ui가 어떻게 나올지에 따라 달라질 거 같은데 일단 api가 따로 있는 쪽으로 수정해보겠습니다~!

shkisme commented on src/main/java/doore/study/application/StudyCommandService.java:
```diff
@@ -0,0 +1,53 @@
+    private final CurriculumItemRepository curriculumItemRepository;
+
+    public void createStudy(final StudyCreateRequest request, final Long teamId) {
+        teamRepository.findById(teamId).orElseThrow(() -> new TeamException(TeamExceptionType.NOT_FOUND_TEAM));
```
왜 팀을 조회하는지, 의도를 나타낼 수 있도록 이름을 가진 메서드로 추출해봐도 좋을 것 같아요!
ex) checkExistTeam()

shkisme commented on src/main/java/doore/study/application/StudyCommandService.java:
```diff
@@ -0,0 +1,53 @@
+
+    public void createStudy(final StudyCreateRequest request, final Long teamId) {
+        teamRepository.findById(teamId).orElseThrow(() -> new TeamException(TeamExceptionType.NOT_FOUND_TEAM));
+        Study study = studyRepository.save(request.toEntityWithoutCurriculum(teamId));
+        List<CurriculumItem> curriculumItems = request.toCurriculumListEntity(study);
```
비즈니스 로직의 많은 부분을 Dto가 들고 있는 것 처럼 보이네요. Dto는 데이터 전송의 역할만을 담당하게 하는 건 어떨까요?

shkisme replied:
시작 날짜가 종료 날짜보다 앞선 날짜가 맞는 지 유효성 검사도 하면 좋을 것 같습니다 🙂

shkisme commented on src/main/java/doore/study/application/StudyCommandService.java:
```diff
@@ -0,0 +1,53 @@
+@RequiredArgsConstructor
+public class StudyCommandService {
+    private final StudyRepository studyRepository;
+    private final TeamRepository teamRepository;
+    private final CurriculumItemRepository curriculumItemRepository;
+
+    public void createStudy(final StudyCreateRequest request, final Long teamId) {
+        teamRepository.findById(teamId).orElseThrow(() -> new TeamException(TeamExceptionType.NOT_FOUND_TEAM));
+        Study study = studyRepository.save(request.toEntityWithoutCurriculum(teamId));
+        List<CurriculumItem> curriculumItems = request.toCurriculumListEntity(study);
+        curriculumItemRepository.saveAll(curriculumItems);
+    }
+
+    public void deleteStudy(Long studyId) {
+        studyRepository.deleteById(studyId);
+    }
```
질문) 존재하지 않는 studyId면 어떻게 되나요?

lcqff replied:
그러면 그냥 아예 무시됩니다! 그래서 예외처리도 따로 안했어욥

02ggang9 replied:
존재하지 않는 ID라고 알려주는건 어떨까용?.?

02ggang9 replied:
생각해보니 아직까지 권한은 생각하지 않기로 했었나요?.? 기억이 잘..

shkisme commented on src/main/java/doore/study/application/StudyCommandService.java:
```diff
@@ -0,0 +1,53 @@
+
+    public void terminateStudy(Long studyId) {
+        Study study = studyRepository.findById(studyId).orElseThrow(() -> new StudyException(NOT_FOUND_STUDY));
+        if (study.isEnded()) {
+            throw new StudyException(ALREADY_TERMINATED_STUDY);
+        }
```
종료된 스터디 -> 종료된 스터디로 변경하면 에러가 발생해야 할까요? 아무 일도 안일어나도 괜찮아 보여서요!

shkisme commented on src/main/java/doore/study/application/dto/request/CurriculumItemsRequest.java:
```diff
@@ -0,0 +1,23 @@
+import jakarta.validation.constraints.NotNull;
+
+public record CurriculumItemsRequest(
+        @NotNull(message = "이름을 입력해주세요.")
+        String name,
+        @NotNull
+        Integer itemOrder,
+        @NotNull
+        Boolean isDeleted
```
어노테이션 간에는 개행이 있으면 가독성이 좋습니다 🙂 그리고, 모든 필드들에 동일하게 에러 메세지를 작성하시면 좋을 것 같아요

shkisme commented on src/main/java/doore/study/application/dto/request/StudyCreateRequest.java:
```diff
@@ -0,0 +1,68 @@
+        @Nullable
+        List<CurriculumItemsRequest> curriculumItems
+) {
+    @Builder
```
빌더로 dto를 생성하고 싶으신 거라면 클래스 단에만 붙여놔도 동작할 것 같습니다!

shkisme commented on src/main/java/doore/study/application/dto/request/StudyCreateRequest.java:
```diff
@@ -0,0 +1,68 @@
+        @NotNull(message = "작물을 골라주세요.")
+        Long cropId,
+
+        @Nullable
+        List<CurriculumItemsRequest> curriculumItems
```
받는 리스트가 아예 null이어도 괜찮을까요? 🤔 커리큘럼이 없다면 빈 리스트를 받는게 맞지 않을까 하는 생각이 듭니다~

lcqff replied:
스터디 생성시에 받아오는 커리큘럼이 null인 경우 비어있는 리스트로 바꿔주는 로직이 다른 메서드에 있긴 했는데 생성자에서 처리하는 편이 나을 것 같아 그렇게 바꿨습니다!
보름님 말씀대로 하긴 했는데 이게 맞는건지 모르겠어서 한번만 더 확인 부탁드립니다😅

shkisme commented on src/main/java/doore/study/domain/Study.java:
```diff
@@ -51,4 +58,45 @@ public class Study extends BaseEntity {
     @OneToMany(mappedBy = "study")
     private List<CurriculumItem> curriculumItems;
 
+
+    public void createCurriculumItems(List<CurriculumItem> newCurriculumItems) {
+        if (curriculumItems == null) {
+            return;
+        }
```
```suggestion
    @OneToMany(mappedBy = "study")
    private final List<CurriculumItem> curriculumItems = new ArrayList<>();


    public void createCurriculumItems(List<CurriculumItem> newCurriculumItems) {
```
컬렉션 필드는 불변을 위해 final로 지정하는 건 어떨까요?
+) 컬렉션 null 체크도 생략하는 건 어떨까요

shkisme commented on src/main/java/doore/study/domain/Study.java:
```diff
@@ -51,4 +58,45 @@ public class Study extends BaseEntity {
+            return;
+        }
+        newCurriculumItems.forEach(newCurriculumItem -> {
+            newCurriculumItem.saveStudyToCurriculum(this);
```
```suggestion
            newCurriculumItem.saveStudy(this);
```
Nit: 메서드 명은 이렇게 해도 설명이 충분한 것 같아요~
연관관계 편의 메서드 사용 좋습니다 👍👍👍

shkisme commented on src/main/java/doore/study/domain/CurriculumItem.java:
```diff
@@ -34,4 +36,15 @@ public class CurriculumItem extends BaseEntity {
     private Study study;
 
+    @Builder
+    public CurriculumItem(String name, Integer itemOrder, Boolean isDeleted, Study study) {
```
생성자를 사용하는 것을 방지하고 싶다면 private 접근 제어자로 변경해주시면 좋겠네요! (Study도 마찬가지)

shkisme commented on src/main/java/doore/study/exception/StudyException.java:
```diff
@@ -0,0 +1,18 @@
+import doore.base.BaseExceptionType;
+
+public class StudyException extends BaseException {
+    StudyExceptionType exceptionType;
```
```suggestion
    private final StudyExceptionType exceptionType;
```

shkisme commented on src/test/java/doore/study/StudyFixture.java:
```diff
@@ -0,0 +1,47 @@
+                .build();
+    }
+
+    public static StudyCreateRequest studyCreateRequest() {
+        return StudyCreateRequest.builder()
+                .name("알고리즘")
+                .description("알고리즘 스터디 입니다.")
+                .startDate(LocalDate.parse("2023-01-01"))
+                .endDate(LocalDate.parse("2024-01-01"))
+                .cropId(1L)
+                .curriculumItems(new ArrayList<CurriculumItemsRequest>())
+                .build();
+    }
```
저희 TestFixture로 만들어두는 객체에 Dto도 포함이었나요?!
Dto는 재사용성이 별로 없다고 생각해서 그때그때 만들어서 사용해도 괜찮지 않은가 싶습니다~

shkisme commented on src/test/java/doore/study/StudyFixture.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.ArrayList;
+
+public class StudyFixture {
+    public static Study algorithm_study() {
```
snake case 발견?! 🐍

shkisme reviewed (changes requested):
꼼꼼히 리뷰하다보니 많이 남기게 되었네요...ㅎㅎㅎㅎ
코멘트 확인 부탁드립니다~!

lcqff committed:
refactor: 스터디 커리큘럼은 null이 아니도록 한다.

- 스터디의 curriculum 필드를 final + 비어있는 리스트로 초기화
- 스터디 생성시 null 입력은 비어있는 리스트로 받아온다.

#14

lcqff committed:
refactor: 스터디 수정의 HTTP 메서드를 PUT으로 변경

# 14

lcqff committed:
refactor: validated 어노테이션 추가

#14

lcqff committed:
refactor: Controller 메서드는 ResponseEntity를 반환하도록 한다.

#14

lcqff committed:
feat: 스터디 상태 변경 api 작성

#14

lcqff committed:
refactor: 존재하는 팀을 확인하는 메서드 분리

#14

lcqff committed:
refactor: dto의 비즈니스 로직을 service로 이동

#14

lcqff committed:
feat: 스터디 종료일이 시작일보다 앞설 경우 예외처리

#14

lcqff committed:
refactor: 종료된 스터디를 종료하는 경우 예외처리 삭제

#14

lcqff committed:
refactor: 어노테이션의 에러메세지 작성 + 개행추가

#14

lcqff committed:
refactor: 빌더 어노테이션 클래스 상단으로 이동

#14

lcqff committed:
refactor: 불필요한 코드 삭제, 생성자 private 처리, 메서드 명 변경

- 불필요한 코드 삭제
- Builder 적용된 생성자를 private 처리
- saveStudyToCurriculum -> saveStudy

#14

lcqff committed:
refactor: 예외 타입 private final 처리

#14

lcqff committed:
refactor: dto 픽스쳐 삭제, 스네이크 케이스 수정

#14

JJimini commented on src/main/java/doore/study/api/StudyController.java:
```diff
@@ -0,0 +1,70 @@
+        return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
+    }
+
+    @GetMapping("/studies/{studyId}")
+    @ResponseStatus(HttpStatus.OK)
```
`@ResponseStatus` 를 붙인 메서드와 안붙인 메서드의 차이가 궁금합니다~!

lcqff replied:
어머.. 삭제하는걸 깜빡했습니다! 좋은 지적...

JJimini commented on src/main/java/doore/study/application/dto/request/CurriculumItemsRequest.java:
```diff
@@ -0,0 +1,17 @@
+import doore.study.domain.Study;
+import jakarta.validation.constraints.NotNull;
+
+public record CurriculumItemsRequest(
+        @NotNull(message = "이름을 입력해주세요.")
+        String name,
+
+        @NotNull(message = "커리큘럼 순서를 입력해주세요.")
+        Integer itemOrder,
+
+        @NotNull(message = "커리큘럼 삭제여부를 입력해주세요.")
+        Boolean isDeleted
+) {
+}
```
음.. 이 부분은 제 파트인 것 같아요~! 제가 이미 만들어둔 `CurriculumItemRequest`가 있고 팜님의 `Request`와 달라서 충돌이 날 것 같네요~
저는 개인적으로 커리큘럼 순서는 `Request`로 받지 않고 `service`에서 순서대로 넣어주고 있으며, 삭제 여부도 받지 않고 있습니다! `Request`를 구성하신 다른 의도가 있는지 궁금합니다~

lcqff replied:
스터디 생성 dto에 커리큘럼이 포함되다보니 일단 커리큘럼 dto를 만들어두었습니다! 딱히 의도는 없어서 미나님 걸로 교체하면 될거같아요😄
그럼 머지하기 전에 커리큘럼 dto(request,response)삭제하면 될까요?

JJimini replied:
이 부분 나중에 회의 때 말씀드리겠습니당!

JJimini commented on src/main/java/doore/study/application/dto/request/StudyUpdateRequest.java:
```diff
@@ -0,0 +1,37 @@
+        @JsonFormat(pattern = "yyyy-MM-dd")
+        LocalDate endDate,
+
+        @NotNull(message = "현재 상태를 입력해주세요.")
+        StudyStatus status
```
커리큘럼을 생성할 때는 상태를 입력 받지 않고, 수정할 때만 받나요??

lcqff replied:
넵! 원래 입력받는 쪽으로 코드를 짰는데 금요일날 회의하면서 스터디 생성할 때 상태를 진행전/진행중/종료 중 하나로 고를 필요 있냐는 의견이 있어서 반영했습니다!

JJimini replied:
이 부분 나중에 회의때 말씀드리겠습니당!

JJimini commented on src/main/java/doore/study/application/dto/response/CurriculumItemResponse.java:
```diff
@@ -0,0 +1,16 @@
+package doore.study.application.dto.response;
+
+import doore.study.domain.CurriculumItem;
+
+public record CurriculumItemResponse(
```
이 부분도 윗 코멘트와 같이 겹치는 부분 같아 보입니다~!

JJimini commented on src/test/java/doore/helper/IntegrationTest.java:
```diff
@@ -38,4 +42,23 @@ protected ResultActions callPostApi(final String url, final Object value) throws
                 .content(objectMapper.writeValueAsString(value)));
     }
+
+    protected ResultActions callDeleteApi(final String url) throws Exception {
+        return mockMvc.perform(delete(url));
+    }
```
만들까 고민했었는데 먼저 만들어주셨군요!👍👍

JJimini commented on src/test/java/doore/restdocs/docs/StudyApiDocsTest.java:
```diff
@@ -0,0 +1,124 @@
+    @MockBean
+    protected StudyRepository studyRepository;
+
+    @Test
+    @DisplayName("스터디를 생성한다.")
```
displayname에 성공&삭제 여부 적는 부분은 적는 방향으로 결정되면, 수정하면 좋을 것 같아요!

JJimini commented on src/main/java/doore/study/application/StudyCommandService.java:
```diff
@@ -0,0 +1,102 @@
+        curriculumItemRepository.saveAll(curriculumItems);
+    }
+
+    private void checkEndDateValid(LocalDate startDate, LocalDate endDate) {
+        if (endDate != null && startDate.isAfter(endDate)) {
+            throw new StudyException(INVALID_ENDDATE);
+        }
+    }
```
그렇다면 생성할 때, 스터디의 `endDate`는 `null`값이어야 한다는 것인가요??

lcqff replied:
endDate를 null로 넘겨주거나, 혹은 endDate를 입력했으나 startDate.isAfter(endDate)을 만족하지 않았을시 패스될거 같습니다!

JJimini replied:
헉 제가 `||` 이랑 헷갈려서 질문했네요..ㅠ

JJimini reviewed (changes requested):
수고하셨습니다~~👍 아무래도 저랑 겹치는 부분이 있다 보니, 확인해주셔야 할 부분이 있는 것 같아요!! 확인 부탁드립니당!

02ggang9 commented on src/main/java/doore/study/api/StudyController.java:
```diff
@@ -0,0 +1,70 @@
+        return ResponseEntity.ok(studyDetailResponse);
+    }
+
+    @PutMapping("/studies/{studyId}")
+    @ResponseStatus(HttpStatus.OK)
+    public ResponseEntity<Void> updateStudy(@Valid @RequestBody StudyUpdateRequest studyUpdateRequest, @PathVariable Long studyId) {
+        studyCommandService.updateStudy(studyUpdateRequest, studyId);
+        return ResponseEntity.status(HttpStatus.OK).build();
+    }
```
메서드의 return 값으로 200값을 내려주는데 @ResponseStatus 어노테이션을 붙여주신 이유가 궁금합니다~!

이번 프로젝트에서 처음 보는데 궁금하네용!

lcqff replied:
원래 responseEntity 없이 ResponseStatus로 응답코드를 보내줬는데 ResponseEntity로 통일하자는 피드백 받고 수정하면서 어노테이션을 지우는걸 깜빡했습니다,,, 하하

02ggang9 commented on src/main/java/doore/study/api/StudyController.java:
```diff
@@ -0,0 +1,70 @@
+
+    @PatchMapping("/studies/{studyId}/status")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public ResponseEntity<Void> changeStudyStatus(@RequestParam String status, @PathVariable Long studyId) {
```
밑에 service 로직의 메서드는 final 을 붙여주셔가지고! 혹시 싶어서 코멘트 답니다~!

02ggang9 commented on src/main/java/doore/study/application/StudyCommandService.java:
```diff
@@ -0,0 +1,102 @@
+        teamRepository.findById(teamId).orElseThrow(() -> new TeamException(TeamExceptionType.NOT_FOUND_TEAM));
+    }
+
+    public Study toStudyWithoutCurriculum(StudyCreateRequest request, Long teamId) {
+        return Study.builder()
+                .name(request.name())
+                .description(request.description())
+                .startDate(request.startDate())
+                .endDate(request.endDate())
+                .status(UPCOMING)
+                .isDeleted(false)
+                .teamId(teamId)
+                .cropId(request.cropId())
+                .build();
+    }
```
사용되는 메서드 바로 밑에 위치하면 좀 더 읽기 편할 것 같습니다~

02ggang9 commented on src/main/java/doore/study/application/StudyCommandService.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    public List<CurriculumItem> toCurriculumList(StudyCreateRequest request, Study study) {
+        return request.curriculumItems().stream()
+                .map(curriculumItemsRequest -> extractCurriculumItemFromStudy(curriculumItemsRequest, study))
+                .toList();
```
👍👍

02ggang9 commented on src/main/java/doore/study/application/dto/request/CurriculumItemsRequest.java:
```diff
@@ -0,0 +1,17 @@
+        @NotNull(message = "커리큘럼 순서를 입력해주세요.")
+        Integer itemOrder,
+
+        @NotNull(message = "커리큘럼 삭제여부를 입력해주세요.")
```
예외 메시지 남겨주는거 👍👍

02ggang9 commented on src/main/java/doore/study/domain/Study.java:
```diff
@@ -49,6 +54,43 @@ public class Study extends BaseEntity {
+    }
+
+    @Builder
+    private Study(String name, String description, LocalDate startDate, LocalDate endDate, StudyStatus status,
+                 Boolean isDeleted, Long teamId, Long cropId, List<CurriculumItem> curriculumItems) {
+        this.name = name;
+        this.description = description;
+        this.startDate = startDate;
+        this.endDate = endDate;
+        this.status = status;
+        this.isDeleted = isDeleted;
+        this.teamId = teamId;
+        this.cropId = cropId;
+        if (curriculumItems != null) {
+            this.curriculumItems.addAll(curriculumItems);
+        }
+    }
```
정말 사소하지만 생성자 밑에 update나 createCurriculumItems가 있는건 어떨까용?.?

02ggang9 commented on src/test/java/doore/study/api/StudyControllerTest.java:
```diff
@@ -0,0 +1,107 @@
+
+        @ParameterizedTest
+        @DisplayName("필수값이 입력되지 않은 경우 스터디 생성에 실패한다.")
+        @CsvSource({
+                ", 알고리즘 스터디입니다., 2022-01-01, 2022-01-05, 1",
+                "알고리즘, , 2022-01-01, 2022-01-05, 1",
+                "알고리즘, 알고리즘 스터디입니다., , 2022-01-05, 1",
+                "알고리즘, 알고리즘 스터디입니다., 2022-01-01, 2022-01-05, "
+        })
```
얘는 뭘 해주나용?.? 처음보네요!

lcqff replied:
문자열 안에 있는 데이터를 ,를 기준으로 차례대로 파라미터 안에 넣어줍니다! 그래서 5가지 파라미터를 각 문자열 대로 4번 테스트하게 됩니다
null인 데이터가 잘 처리되는지 확인하려고 사용했어요!

02ggang9 commented on src/test/java/doore/study/StudyFixture.java:
```diff
@@ -0,0 +1,26 @@
+
+public class StudyFixture {
+    public static Study algorithmStudy() {
+        return Study.builder()
+                .name("알고리즘")
+                .description("알고리즘 스터디 입니다.")
+                .startDate(LocalDate.parse("2023-01-01"))
+                .endDate(LocalDate.parse("2024-01-01"))
+                .teamId(1L)
+                .status(StudyStatus.IN_PROGRESS)
+                .isDeleted(false)
+                .cropId(1L)
+                .curriculumItems(new ArrayList<CurriculumItem>())
+                .build();
+    }
```
https://github.com/BDD-CLUB/01-doo-re-back/pull/50#discussion_r1467296255

위의 리뷰에서 의견을 냈었는데 @lcqff 님은 어떻게 생각하시는지 궁금하네요!

lcqff replied:
대표적으로 많이 쓰이는 객체는 현재 방식대로 미리 생성해두는게 아니었나요? 
일단 디스커션에서 논의했던대로 빌더패턴으로 수정해보려했는데 빌더 패턴 사용시 실제 사용법이 어떻게 되는지 모르겠어서 일단 논의된 이후 결정된 방향으로 수정하겠습니다! ㅜㅜ

02ggang9 commented on src/test/java/doore/study/domain/StudyTest.java:
```diff
@@ -0,0 +1,52 @@
+    @DisplayName("커리큘럼을 생성할 수 있다.")
+    public void 커리큘럼을_생성할_수_있다_성공() {
+        final Study study = algorithmStudy();
+        CurriculumItem curriculumItem1 = CurriculumItem.builder()
```
단순 1, 2 이런 변수명 보다는 otherCurriculumItem 이런 변수명은 어떤가요?

저도 과거 이런 피드백을 받았던 기억이 나서 PR들을 뒤져봤는데 못 찾았네용 ㅠㅠ..레퍼런스 없는 점 죄송합니닷

02ggang9 reviewed (approved):
수고하셨습니다~~! 👍👍 혹시 리뷰 반영되면 다시한번 requst 걸어주실 수 있을까용?

lcqff committed:
refactor: ResponseStatus 삭제

#14

lcqff committed:
refactor: curriculumItemsRequest 수정

- ItemOrder, isDeleted 삭제

#14

lcqff committed:
feat: 스터디 삭제시 존재하는 스터디가 아닌 경우 예외처리

#14

lcqff committed:
refactor: 사용하는 코드 아래로 메서드 이동

#14

lcqff committed:
refactor: 생성자를 제일 아래에 위치

#14

lcqff committed:
refactor: 변수명을 명시적으로 한다.

shkisme reviewed (approved):
수고 많으셨습니다~!!
